### ISSUE TITLE
Mac OS support

### ISSUE THREAD

**Issue by davidlattimore** [google/evcxr#2]
In the unlikely event that it already works on Mac OS, please let me know.

If you'd like to work on this, feel free to reach out with any questions. I'm not really sure what will be involved, since I don't know what the problems will be.

**Comment by nbigaouette** [comment]
Thanks for this crate! Was curious if I could blindly try this on macOS. Just tried the repl on High Sierra (10.13.6 (17G65)). Compiles, installs ans launches fine, even though I expected a failure due to missing dependency (zqm?).

Unfortunately I can't do anything with it:
```
> evcxr
Welcome to evcxr. For help, type .help
# .help
  ^ expected expression
expected expression, found `.`
#
```

```
> evcxr
Welcome to evcxr. For help, type .help
# let a = 1.0;
While processing instruction `Ok("LOAD_AND_RUN /var/folders/4y/75_ckd416vj53qlqppnp65fh0000gn/T/.tmppVl6ZA/target/debug/deps/user_code_0.dylib run_user_code_0")`, got error: JustMessage("dlopen(/var/folders/4y/75_ckd416vj53qlqppnp65fh0000gn/T/.tmppVl6ZA/target/debug/deps/user_code_0.dylib, 2): image not found")
Child process terminated with status: exit code: 99
#
```

**Comment by nbigaouette** [comment]
I just checked the target folder, and it simply seems like a filename issue. I have the file `libuser_code_0.dylib` and not `user_code_0.dylib`.

**Comment by davidlattimore** [comment]
Thanks for checking this out. The first problem is fixed by https://github.com/google/evcxr/pull/9 - it should have read ":help" instead of ".help".

Regarding actually running on Mac, with evcxr still running and after getting an error like the above, can you have a look in the directory that it's trying to load the dylib from to see if there's a file with a similar name? The directory that would read something like .../target/debug/deps/

**Comment by davidlattimore** [comment]
Ah, cool. The code is [here](https://github.com/google/evcxr/blob/master/evcxr/src/module.rs#L26) if you'd like to try fixing it, otherwise I'll do so this evening (Sydney time)

**Comment by Carreau** [comment]
So the repl same issue with filenames. 

I had slightly more issues to compile things that use ZMQ, but that seen more a ZMQ crate issue. 
here was my quick workaround for any one else having the issue:

```
 ~/dev/evcxr[master ✓] $ cargo build
   Compiling zmq v0.8.2
error: failed to run custom build command for `zmq v0.8.2`
process didn't exit successfully: $HOME/dev/evcxr/target/debug/build/zmq-915ff4fa4740b2f7/build-script-build` (signal: 6, SIGABRT: process abort signal)
--- stderr
dyld: Library not loaded: @rpath/libzmq.5.dylib
  Referenced from: $HOME/dev/evcxr/target/debug/build/zmq-915ff4fa4740b2f7/build-script-build
  Reason: image not found

 ~/dev/evcxr[master ✓] $
 ~/dev/evcxr[master ✓] $ otool -L $HOME/dev/evcxr/target/debug/build/zmq-915ff4fa4740b2f7/build-script-build
$HOME/dev/evcxr/target/debug/build/zmq-915ff4fa4740b2f7/build-script-build:
	@rpath/libzmq.5.dylib (compatibility version 7.0.0, current version 7.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
 ~/dev/evcxr[master ✓] $ install_name_tool -change @rpath/libzmq.5.dylib $HOME/anaconda/lib/libzmq.5.dylib $HOME/dev/evcxr/target/debug/build/zmq-915ff4fa4740b2f7/build-script-build
 ~/dev/evcxr[master ✓] $ otool -L $HOME/dev/evcxr/target/debug/build/zmq-915ff4fa4740b2f7/build-script-build
$HOME/dev/evcxr/target/debug/build/zmq-915ff4fa4740b2f7/build-script-build:
	$HOME/anaconda/lib/libzmq.5.dylib (compatibility version 7.0.0, current version 7.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
 ~/dev/evcxr[master ✓] $ cargo build
   Compiling zmq v0.8.2
   Compiling evcxr_jupyter v0.1.1 (file:///$HOME/dev/evcxr/evcxr_jupyter)
    Finished dev [unoptimized + debuginfo] target(s) in 4.94s
 ~/dev/evcxr[master ✓] $
```

Same for `evcxr_jupyter` not sure it is just my system I'm still on sierra.

**Comment by Carreau** [comment]
Seem to partially work with the modification suggested, a least for simple assignments, defining a struct and displaying it fails _the first time_


```
#[derive(Debug)]
pub struct Foo{
    pub a:i32, 
    pub b:i32
}
```
```
Foo {a:1, b:1}
---
While processing instruction `Ok("LOAD_AND_RUN /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpylFEtc/target/debug/deps/libuser_code_3.dylib run_user_code_3")`, got error: 
JustMessage("dlopen(/var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpylFEtc/target/debug/deps/libuser_code_3.dylib, 2): Symbol not found: __ZN53_$LT$user_code_2..Foo$u20$as$u20$core..fmt..Debug$GT$3fmt17h47d6626f41d578efE
Referenced from: /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpylFEtc/target/debug/deps/libuser_code_3.dylib (which was built for Mac OS X 10.12)
Expected in: /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpylFEtc/target/debug/deps/libuser_code_2.dylib
in /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpylFEtc/target/debug/deps/libuser_code_3.dylib")
```

But a second time works fine.
```
Foo {a:1, b:1}
---
Foo {a:1, b:1} 
```

**Comment by Carreau** [comment]
> You could try disabling that and see if it makes a difference.

It does not appear to do any difference.

**Comment by nbigaouette** [comment]
Probably the same issue as https://github.com/rust-lang/cargo/issues/1970

**Comment by kbarros** [comment]
Worked for me after executing

```mv /Users/$USER/Library/Application\ Support/jupyter/kernels/rust /Users/$USER/Library/Jupyter/kernels/```

In other words, it seems like the install directory should be changed? `jupyter --version` returns 4.4.0 .

**Comment by vegai** [comment]
Seems to work fine for many basic cases on my 2017 Macbook Pro running Mojave:

```
» ./evcxr
Welcome to evcxr. For help, type :help
>> :timing
Timing: true
>> 1+2
3
Took 1435ms
>> println!("Foo {}", "bar")
Foo bar
()
Took 699ms
>> :help
:vars             List bound variables and their types
:opt              Toggle optimization
:explain          Print explanation of last error
:clear            Clear all state, keeping compilation cache
:dep              Add dependency. e.g. :dep regex = "1.0"

Mostly for development / debugging purposes:
:last_compile_dir Print the directory in which we last compiled
:timing           Toggle printing of how long evaluations take
:last_error_json  Print the last compilation error as JSON (for debugging)
:internal_debug   Toggle various internal debugging code
>> :dep regex = "1.0"
>> extern crate regex;

Took 1183ms
>> use regex::Regex;

Took 1152ms
>> let re = Regex::new(r"^\d{4}-\d{2}-\d{2}$").unwrap();

Took 3407ms
>> assert!(re.is_match("2014-01-01"));

Took 1951ms
>> 
```

Extremely cool stuff!

**Comment by davidlattimore** [comment]
Thanks for that report! Could someone with a Mac check if `cargo test --all` passes? If it does, I'll look at adding Mac to the Travis config.

**Comment by Carreau** [comment]
Seem there are some failures here (Sierra, nightly-2018-07-31-x86_64-apple-darwin, I should try on a newer version)


<details>
<pre>
$ cargo test --all
   Compiling num-traits v0.2.6
   Compiling num-integer v0.1.39
   Compiling evcxr_runtime v1.0.0 (file:///Users/bussonniermatthias/dev/evcxr/evcxr_runtime)
   Compiling time v0.1.40
   Compiling evcxr v0.2.0 (file:///Users/bussonniermatthias/dev/evcxr/evcxr)
   Compiling chrono v0.4.6
   Compiling evcxr_repl v0.2.0 (file:///Users/bussonniermatthias/dev/evcxr/evcxr_repl)
   Compiling evcxr_jupyter v0.2.0 (file:///Users/bussonniermatthias/dev/evcxr/evcxr_jupyter)
    Finished dev [unoptimized + debuginfo] target(s) in 22.28s
     Running target/debug/deps/evcxr-622620c695e8d0e4

running 11 tests
test crate_config::tests::make_paths_absolute ... ignored
test code_block::test::basic_usage ... ok
test idents::tests::destructure_tuple ... ok
test cargo_metadata::tests::test_library_names_from_metadata ... ok
test idents::tests::destructure_struct ... ok
test idents::tests::destructure_tuple_struct ... ok
test idents::tests::simple_variable ... ok
test statement_splitter::test::statement_ends_with_brace ... ok
test statement_splitter::test::single_line ... ok
test statement_splitter::test::multiple_lines ... ok
test statement_splitter::test::else_statements ... ok

test result: ok. 10 passed; 0 failed; 1 ignored; 0 measured; 0 filtered out

     Running target/debug/deps/testing_runtime-3d51b254d2b01771

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running target/debug/deps/integration_tests-fa3763de556520ff

running 16 tests
test error_from_macro_expansion ... ok
While processing instruction `Ok("LOAD_AND_RUN /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpnhfh5q/target/debug/deps/libuser_code_1.dylib run_user_code_1")`, got error: JustMessage("dlopen(/var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpnhfh5q/target/debug/deps/libuser_code_1.dylib, 2): Symbol not found: __ZN11user_code_03bar17h86acb07473e22ba8E\n  Referenced from: /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpnhfh5q/target/debug/deps/libuser_code_1.dylib (which was built for Mac OS X 10.12)\n  Expected in: /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpnhfh5q/target/debug/deps/libuser_code_0.dylib\n in /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpnhfh5q/target/debug/deps/libuser_code_1.dylib")
test define_then_call_function ... FAILED
test continue_execution_after_bad_use_statement ... ok
test multiple_identical_use_statements ... ok
test moved_value ... ok
test function_panics ... ok
test non_concrete_types ... ok
test printing ... ok
While processing instruction `Ok("LOAD_AND_RUN /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpItzVZN/target/debug/deps/libuser_code_4.dylib run_user_code_4")`, got error: JustMessage("dlopen(/var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpItzVZN/target/debug/deps/libuser_code_4.dylib, 2): Symbol not found: __ZN11user_code_03Foo9get_value17h26cc20b6e887ba44E\n  Referenced from: /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpItzVZN/target/debug/deps/libuser_code_4.dylib (which was built for Mac OS X 10.12)\n  Expected in: /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpItzVZN/target/debug/deps/libuser_code_0.dylib\n in /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpItzVZN/target/debug/deps/libuser_code_4.dylib")
test redefine_type_reference_old_var ... FAILED
test save_and_restore_variables ... ok
test rc_refcell_etc ... ok
test single_statement ... ok
test statement_and_expression ... ok
While processing instruction `Ok("LOAD_AND_RUN /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpNy6z3a/target/debug/deps/libuser_code_1.dylib run_user_code_1")`, got error: JustMessage("dlopen(/var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpNy6z3a/target/debug/deps/libuser_code_1.dylib, 2): Symbol not found: __ZN11user_code_08init_foo17h1a74787c67c3adebE\n  Referenced from: /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpNy6z3a/target/debug/deps/libuser_code_1.dylib (which was built for Mac OS X 10.12)\n  Expected in: /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpNy6z3a/target/debug/deps/libuser_code_0.dylib\n in /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmpNy6z3a/target/debug/deps/libuser_code_1.dylib")
test tls_implementing_drop ... FAILED
While processing instruction `Ok("LOAD_AND_RUN /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmp5WdAS8/target/debug/deps/libuser_code_2.dylib run_user_code_2")`, got error: JustMessage("dlopen(/var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmp5WdAS8/target/debug/deps/libuser_code_2.dylib, 2): Symbol not found: __ZN55_$LT$user_code_0..Point$u20$as$u20$core..fmt..Debug$GT$3fmt17h24b7c19f68bb4a71E\n  Referenced from: /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmp5WdAS8/target/debug/deps/libuser_code_2.dylib (which was built for Mac OS X 10.12)\n  Expected in: /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmp5WdAS8/target/debug/deps/libuser_code_0.dylib\n in /var/folders/ld/jy0c_5d91rl_s3jqh74tmqn00000gn/T/.tmp5WdAS8/target/debug/deps/libuser_code_2.dylib")
test struct_type_inference ... FAILED
test abort_and_restart ... test abort_and_restart has been running for over 60 seconds 
^C
</pre>
</details>

**Comment by untitaker** [comment]
I don't know what happened since this issue has been opened, but this crate works perfectly for me on Mojave.

**Comment by davidlattimore** [comment]
You're right. It was fixed in 0.3.3 by Daniel Griffen. Thanks for the prompt, and I'm glad it's working well for you.
